src/utils/events/index.ts

```diff
@@ -194,57 +194,62 @@ function mergeAbiEvents(target: any, source: any): Object {
  * ```
  */
 export function parseEvents(
-  providerReceivedEvents: RPC.Event[],
+  providerReceivedEvents: RPC.EmittedEvent[] | RPC.Event[],
```

Comment:
Why `RPC.EmittedEvent[] | RPC.Event[]`?
Isn't it only `RPC.EmittedEvent[]`?
In which case are  block_hash, block_number & transaction_hash not included?

Reply:
Block/tx info is not included in a case where you first fetch transaction receipt with `starknet_getTransactionReceipt`. In those kinds of reponses `RPC.Event` struct is returned (see: [this](https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L2661)). On the other hand when fetching data with `starknet_fetchEvents` returned struct is `RPC.EmittedEvents`.

So I thought it would be nice to leave that as is in case someone had code like this
```js
events.parseEvents(receipts.map(t => t.events),...)
```

Reply:
I think that the root cause of this wrong input type of `events.parseEvents()` is due to `Contract.parseEvent()`. This method is calling `events.parseEvents()` with `RPC.event `format ; here is the root cause of all this mess.
I think that you have to modify `Contract.parseEvent()` : 
- if the event is included in a block, this type has to be sent to `events.parseEvents()` :
```
type EMITTED_EVENT = EVENT & {
  block_hash: BLOCK_HASH;
  block_number: BLOCK_NUMBER;
  transaction_hash: TXN_HASH;
};
```
- if the event is in a PENDING transaction, this type has to be sent to `events.parseEvents()` :
```
type EMITTED_EVENT = EVENT & {
  transaction_hash: TXN_HASH;
};
```
`events.parseEvents()` has to accept both types as input.

`RPC.Event[]` type was here only to handle `Contract.parseEvent()` ; it's useless for users. They uses `EVENTS_CHUNK`type (result of `RpcProvider.getEvents()`), that includes `EmittedEvent` type. Never `RPC.Events`.

Reply:
Not sure how pending transaction have anything to do with this. AFAIK events are created and emitted during transaction execution. So there is always tx hash, block hash and block number corresponding to the event. They are missing from events array of the recipiet because that is always available in parent (receipt) object.

Reply:
If PENDING is not appropriate, then we have still a problem in Contract.parseEvents() : 
https://github.com/starknet-io/starknet.js/blob/47e52cf39a71bf99188edc4991b002018e296504/src/contract/default.ts#L333
`InvokeTransactionReceiptResponse` type is not the good one, because it's made of `INVOKE_TXN_RECEIPT | PENDING_INVOKE_TXN_RECEIPT`
And in this case, only `EMITTED_EVENT`  should be transfered from `Contract.parseEvent()` to `events.parseEvents()` (not `RPC.event`).
Comment about uselessness of `RPC.event` as input of `events.parseEvents()` remains valid.

Reply:
So if I understood correctly, you would like `Contract.parseEvent()` to change and look like this?
```ts
  public parseEvents(receipt: GetTransactionReceiptResponse): ParsedEvents {
    return parseRawEvents(
      (receipt as InvokeTransactionReceiptResponse).events?.
      map(event => { return {
        block_hash: (receipt as INVOKE_TXN_RECEIPT).block_hash,
        block_number: (receipt as INVOKE_TXN_RECEIPT).block_number,
        transaction_hash: (receipt as INVOKE_TXN_RECEIPT).transaction_hash,
        ...event
      }})
      .filter(
        (event) => cleanHex(event.from_address) === cleanHex(this.address),
        []
      ) || [],
      this.events,
      this.structs,
      CallData.getAbiEnum(this.abi)
    );
  }
  ```

Reply:
Yes, this way.
I tested successfully this :
**account/default.ts**
```typescript
public parseEvents(receipt: GetTransactionReceiptResponse): ParsedEvents {
let parsed: ParsedEvents;
receipt.match({
  success: (txR: SuccessfulTransactionReceiptResponse) => {
    const emittedEvents =
      (txR as InvokeTransactionReceiptResponse).events
        ?.map((event) => {
          return {
            block_hash: (txR as INVOKE_TXN_RECEIPT).block_hash,
            block_number: (txR as INVOKE_TXN_RECEIPT).block_number,
            transaction_hash: (txR as INVOKE_TXN_RECEIPT).transaction_hash,
            ...event,
          };
        })
        .filter((event) => cleanHex(event.from_address) === cleanHex(this.address), []) || [];
    parsed = parseRawEvents(
      emittedEvents,
      this.events,
      this.structs,
      CallData.getAbiEnum(this.abi)
    );
  },
  _: () => {
    throw Error('This transaction was not successful.');
  },
});
return parsed!;
}
```

**events/index.ts**
```typescript
export function parseEvents(
  providerReceivedEvents: RPC.EmittedEvent[],
  abiEvents: AbiEvents,
  abiStructs: AbiStructs,
  abiEnums: AbiEnums
): ParsedEvents {
...
```